### PR TITLE
Add trafficops api  endpoints to fetch and save uri signing keys.

### PR DESCRIPTION
**Pull request by jrushford**
This PR adds rest endpoints to fetch and store jwk keysets to riak by delivery service for the new ATS cdn uri signing plugin, see issue https://github.com/apache/incubator-trafficcontrol/issues/1362

` {1.3, http.MethodGet, "deliveryservices/{xml-id}/urisignkeys$",`
 
 `{1.3, http.MethodPost, "deliveryservices/{xml-id}/urisignkeys$", `

A new bucket is created in RIAK for this purpose.  Execute this command to create the bucket:

$ riak-admin security grant riak_kv.get,riak_kv.put,riak_kv.delete on default cdn_uri_sig_keys to keysusers

The api adds two vendored  libraries as this api uses the native Riak protobuffer endpoints.  
        vendor/github.com/go-jwx
        vendor/github.com/basho/riak-go-client

Files changed or added, not including the vendored library files are:

traffic_ops/app/conf/development/riak.conf
traffic_ops/app/conf/integration/riak.conf
traffic_ops/app/conf/production/riak.conf
traffic_ops/app/conf/test/riak.conf
traffic_ops/traffic_ops_golang/config.go
traffic_ops/traffic_ops_golang/perlconfig.go
traffic_ops/traffic_ops_golang/perlconfig_test.go
traffic_ops/traffic_ops_golang/riak.go
traffic_ops/traffic_ops_golang/routes.go
traffic_ops/traffic_ops_golang/traffic_ops_golang.go
traffic_ops/traffic_ops_golang/wrappers.go

### CODE REVIEW

File: traffic_ops/app/conf/production/riak.conf [hunk 1]
```diff
@@ -1,5 +1,7 @@
-
 {
 	"user": "riakuser",
-	"password": "password"
+	"password": "password",
+	"tlsConfig": {
+		"insecureSkipVerify": true
```
Comment (PSUdaemon) [inline traffic_ops/app/conf/production/riak.conf]: Do we need this in prod?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/397/

**Comment by dneuman64** [comment]
can you update the docs with your new riak-admin command?
https://github.com/apache/incubator-trafficcontrol/blob/master/docs/source/admin/traffic_vault.rst

File: traffic_ops/traffic_ops_golang/riak.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
+package main
+
+import (
+	//"database/sql"
```
Comment (PSUdaemon) [inline traffic_ops/traffic_ops_golang/riak.go]: Delete?

**Comment by jrushford** [comment]
@dneuman64 - docs are updated.  See 2nd commit.

File: traffic_ops/traffic_ops_golang/wrappers.go [hunk 3]
```diff
@@ -186,7 +186,7 @@ func gzipIfAccepts(r *http.Request, w http.ResponseWriter, b []byte) ([]byte, er
 	if len(b) == 0 || !acceptsGzip(r) {
 		return b, nil
 	}
-	w.Header().Set(api.ContentType, api.Gzip)
+	w.Header().Set("Content-Encoding", api.Gzip)
```
Comment (PSUdaemon) [inline traffic_ops/traffic_ops_golang/wrappers.go]: What's this little guy about? Seems unrelated to the rest of the PR.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/398/

File: traffic_ops/traffic_ops_golang/wrappers.go (hunk 3 above)
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/wrappers.go]: Fixes https://github.com/apache/incubator-trafficcontrol/issues/1354
I assume this was an accidental commit, added while debugging to work around the bug in master. Needs to be reverted, because there's already a PR fixing which will cause conflicts https://github.com/apache/incubator-trafficcontrol/pull/1359

Comment (jrushford) [reply traffic_ops/traffic_ops_golang/wrappers.go]: this was a bug we found while testing.  My ContentType header of "application/json" was being overwritten by this setting and the son was being returned gzipped.  Rob spotted this and wrote a bug up on it as the content encoding and not the type should be gzip.

File: traffic_ops/traffic_ops_golang/riak.go (hunk 2 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/riak.go]: missed that, I'll delete it.

File: traffic_ops/app/conf/production/riak.conf (hunk 1 above)
Comment (jrushford) [reply traffic_ops/app/conf/production/riak.conf]: Didn't test in prod and I'm not sure if it would work in prod without it as you would have to set the ServerName in TLSConfig and I think ServerName verification would fail when selecting a RIAK node from a cluster of nodes.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/354/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/355/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/399/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/400/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/356/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/357/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/401/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/358/

**Comment by mitchell852** [comment]
@jrushford - can you create a simple issue (https://github.com/apache/incubator-trafficcontrol/issues) with a short description of what you adding exactly. really, i'm just curious which api endpoints are being added.

and then you can link the issue to this PR

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/406/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/362/

**Comment by jrushford** [comment]
@mitchell852 I updated the PR comments and added info about the api endpoints.  Is there api documentation that needs updating?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/409/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/365/

**Comment by mitchell852** [comment]
@jrushford - yes, you should add these endpoints to the deliveryservice.rst file for version 1.2. thanks.

**Comment by dneuman64** [comment]
@mitchell852 why would he add it to the 1.2 docs when it is a 1.3 route?

File: traffic_ops/traffic_ops_golang/riak.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/apache/incubator-trafficcontrol/lib/go-log"
+	"github.com/apache/incubator-trafficcontrol/traffic_ops/tostructs"
+	"github.com/basho/riak-go-client"
+	"github.com/jmoiron/sqlx"
+	"github.com/lestrrat/go-jwx/jwk"
+	"io/ioutil"
+	"net/http"
+)
+
+const riak_port = 8087
```
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/riak.go]: Go Constants should be named the same as other global symbols, with MixedCase. See https://golang.org/doc/effective_go.html#mixed-caps

File: traffic_ops/traffic_ops_golang/riak.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+		xmlId := p["xml-id"]
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		// validate that the received data is a valid jwk keyset
+		var keys jwk.SymmetricKey
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/riak.go]: This is incorrect. You'll need a custom type that probably looks something like:

    type URISignerKeyset struct {
        jwk.Set
        RenewalKid string `json:"renewal_kid,omitempty"`
    }
    type URISignerConfig map[string]URISignerKeyset

File: traffic_ops/traffic_ops_golang/riak.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+		nodes[i], err = riak.NewNode(nodeOpts)
+		if err != nil {
+			return nil, err
+		}
+		count++
+		i++
+	}
+
+	opts := &riak.ClusterOptions{
+		Nodes: nodes[0 : count-1],
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/riak.go]: It appears `maxNodes` isn't exceptionally large. (Hardcoded to 12?) This subslice doesn't free the extra nodes you allocated, but didn't use. If, like it appears, this isn't a relevant amount of memory and isn't kept for an inordinate amount of time, that's totally fine. But if I'm misreading it, this could cause more memory to be used than you might guess. (The "fix" is to copy the memory, which might be more expensive, so I think this is still probably best.)

File: traffic_ops/traffic_ops_golang/riak.go (hunk 5 above)
Comment (alficles) [reply traffic_ops/traffic_ops_golang/riak.go]: Here's the readme: https://gist.github.com/alficles/a375f64b563993b608e69ec5a75a13be

(I tossed the README in a gist since the actual plugin isn't quite polished for submission to upstream yet.)

File: traffic_ops/traffic_ops_golang/routes.go [hunk 7]
```diff
@@ -56,6 +56,8 @@ func Routes(d ServerData) ([]Route, http.Handler, error) {
 		{1.2, http.MethodGet, "divisions-wip.json$", wrapHeaders(wrapAuthWithData(divisionsHandler(d.DB), d.Insecure, d.TOSecret, rd.PrivLevelStmt, DivisionsPrivLevel))},
 		{1.2, http.MethodGet, "hwinfo-wip$", wrapHeaders(wrapAuthWithData(hwInfoHandler(d.DB), d.Insecure, d.TOSecret, rd.PrivLevelStmt, HWInfoPrivLevel))},
 		{1.2, http.MethodGet, "hwinfo-wip.json$", wrapHeaders(wrapAuthWithData(hwInfoHandler(d.DB), d.Insecure, d.TOSecret, rd.PrivLevelStmt, HWInfoPrivLevel))},
+		{1.3, http.MethodGet, "deliveryservices/{xml-id}/urisignkeys$", wrapHeaders(wrapAuth(urisignkeysHandler(d.DB, d.Config), d.Insecure, d.TOSecret, rd.PrivLevelStmt, HWInfoPrivLevel))},
+		{1.3, http.MethodPost, "deliveryservices/{xml-id}/urisignkeys$", wrapHeaders(wrapAuthWithData(assignDeliveryServiceUriKeysKeysHandler(d.DB, d.Config), d.Insecure, d.TOSecret, rd.PrivLevelStmt, HWInfoPrivLevel))},
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/routes.go]: Should a read-only user be able to read and update signing keys‽ I think key access should be a bit more restricted, but maybe we have it mostly open? Thoughts?

File: traffic_ops/traffic_ops_golang/riak.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+	if len(resp.Values) > 0 {
+		obj = resp.Values[0]
+	} else {
+		return "", fmt.Errorf("no such object")
+	}
+	return string(obj.Value), nil
+}
+
+func assignDeliveryServiceUriKeysKeysHandler(db *sqlx.DB, cfg Config) AuthRegexHandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, p PathParams, username string, privLevel int) {
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/riak.go]: Why are we using `withData` if we don't need differentiation by `privLevel` on data handling?

File: traffic_ops/traffic_ops_golang/riak.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+	"net/http"
+)
+
+const riak_port = 8087
+const cdn_uri_keys_bucket = "cdn_uri_sig_keys" // riak namespace for cdn uri signing keys.
+
+func getStringValue(resp *riak.FetchValueResponse) (string, error) {
+	var obj *riak.Object
+
+	if len(resp.Values) > 0 {
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/riak.go]: You are currently requesting via key, so I _think_ it's impossible to get more than 1 value here. But it feels wrong not to check for the case anyway and error out.

File: traffic_ops/traffic_ops_golang/riak.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/jmoiron/sqlx"
+	"github.com/lestrrat/go-jwx/jwk"
+	"io/ioutil"
+	"net/http"
+)
+
+const riak_port = 8087
+const cdn_uri_keys_bucket = "cdn_uri_sig_keys" // riak namespace for cdn uri signing keys.
+
+func getStringValue(resp *riak.FetchValueResponse) (string, error) {
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/riak.go]: `getStringValue` is maybe a bit generic of a name for a global function that converts a specific library type into a  string. It's only used in one spot as near as I can tell; you might put it in local scope of the function that uses it.

File: traffic_ops/traffic_ops_golang/riak.go (hunk 5 above)
Comment (alficles) [reply traffic_ops/traffic_ops_golang/riak.go]: And an extra thing you could check, if you wanted: Verify that ∃ exactly one `renewal_kid` and that it matches a key present in the file. Also, that every key has a `kid`. (It's legal to have JWKs without them, but we need them for efficient key-selection.)

File: traffic_ops/traffic_ops_golang/riak.go (hunk 9 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/riak.go]: Okay, I agree, I'll change it to ==

File: traffic_ops/traffic_ops_golang/riak.go (hunk 10 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/riak.go]: It may be used with other objects such as the url_sig keys going forward.  Thats why its there.

File: traffic_ops/traffic_ops_golang/riak.go (hunk 8 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/riak.go]: Will need the username and privLevel as I need to work on the tenancy checks.

File: traffic_ops/traffic_ops_golang/riak.go (hunk 5 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/riak.go]: Okay, my understanding was that we would get a JSON web key set, RFC 7517, from NBC.  So, I can make the change with a custom struct and check the details.  So, I presume that NBC will now what to send us right?

File: traffic_ops/traffic_ops_golang/routes.go (hunk 7 above)
Comment (jrushford) [reply traffic_ops/traffic_ops_golang/routes.go]: I intend to add in tenancy checks to verify the user has access.  Was planning to do that next sprint.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/417/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/436/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/392/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/452/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/408/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/463/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/464/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/419/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR-trafficops-test/420/
